### PR TITLE
Make verifiability optional and add batched DLEQ case

### PR DESCRIPTION
**Pull request by alxdavids**
For #15 and #16, big change so I'm happy to iterate on this.

### CODE REVIEW

File: draft-sullivan-cfrg-voprf.md [hunk 1]
```diff
@@ -167,25 +188,44 @@ trusted public key Y = kG. Informally, this is done by presenting a
...
+wants key consistency from P. This property is necessary in some applications,
+e.g., the Privacy Pass protocol {{PrivacyPass}}, wherein this VOPRF is used to
+generate one-time authentication tokens to bypass CAPTCHA challenges. VOPRFs
+have also been used for password-protected secret sharing schemes e.g.
+{{JKKX16}}.
+
+This document introduces an OPRF protocol built in prime-order groups, applying
+to finite fields of prime-order and also elliptic curve (EC) settings. The
+protocol has the option of being extended to a VOPRF with the addition of a NIZK
+proof for proving discrete log equality relations. In the EC setting, we will
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: The end of this sentence seems a bit out of context. Perhaps: "... with the addition of a proof that demonstrates correctness of the computation using a known public key."

File: draft-sullivan-cfrg-voprf.md [hunk 2]
```diff
@@ -167,25 +188,44 @@ trusted public key Y = kG. Informally, this is done by presenting a
...
+refer to the protocol as ECVOPRF. The document describes the protocol, its
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/ECVOPRF/ECOPRF?

File: draft-sullivan-cfrg-voprf.md [hunk 3]
```diff
@@ -167,25 +188,44 @@ trusted public key Y = kG. Informally, this is done by presenting a
...
+  {{voprf}}.
+- Section {{dleq}}: Specify the NIZK discrete logarithm equality (DLEQ)
+  construction used for constructing the VOPRF protocol.
+- Section {{batch}}: Specifies how the DLEQ proof mechanism can be batched for
+  multiple VOPRF invocations, and how this changes the protocol execution.
+- Section {{ecinstantiation}}: Considers explicit instantiations of the protocol
+  in the elliptic curve setting.
+- Section {{sec}}: Discusses the security considerations for the OPRF and VOPRF
+  protocol.
+- Section {{apps}}: Discusses some existing applications of the OPRF/VOPRF
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/OPRF\/VOPRF protocol/OPRF and VOPRF protocols

File: draft-sullivan-cfrg-voprf.md [hunk 4]
```diff
@@ -208,11 +248,12 @@ interpreted as described in {{RFC2119}}.
 
 # Background {#background}
 
-VOPRFs are functionally related to RSA-based blind signature schemes, e.g.,
-{{ChaumBlindSignature}}. Such a scheme works as follows. Let m be a message to
-be signed by a server. It is assumed to be a member of the RSA group. Also, let
-N be the RSA modulus, and e and d be the public and private keys, respectively.
-A prover P and verifier V engage in the following protocol given input m.
+OPRFs are functionally related to RSA-based blind signature schemes, e.g.,
+{{ChaumBlindSignature}}. Broadly, a blind signature scheme works as follows. Let
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/Broadly/Briefly?

File: draft-sullivan-cfrg-voprf.md [hunk 5]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
    (s')^(r^-1) (mod N).
 
 By the properties of RSA, s is clearly a valid signature for m. OPRF protocols
-are the symmetric equivalent to blind signatures in the same way that PRFs are
-the symmetric equivalent traditional digital signatures. This is discussed more
-in the following section.
+can be used to provide a symmetric equivalent to blind signatures. Essentially
+the client learns PRF(k,x) for some input x of their choice, from a server that
+holds K. Since the security of an OPRF means that x is hidden in the
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/K/k?

File: draft-sullivan-cfrg-voprf.md [hunk 6]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
    (s')^(r^-1) (mod N).
 
 By the properties of RSA, s is clearly a valid signature for m. OPRF protocols
-are the symmetric equivalent to blind signatures in the same way that PRFs are
-the symmetric equivalent traditional digital signatures. This is discussed more
-in the following section.
+can be used to provide a symmetric equivalent to blind signatures. Essentially
+the client learns PRF(k,x) for some input x of their choice, from a server that
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Perhaps we should assign `PRF(k, x)` a value, e.g., `y = PRF(k, x)`?

File: draft-sullivan-cfrg-voprf.md [hunk 7]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
...
+holds K. Since the security of an OPRF means that x is hidden in the
+interaction, then the client can later reveal x to the server along with
+PRF(k,x).
+
+The server can verify that PRF(k,x) is computed correctly by recomputing the PRF
+on x using K. In doing so, the client provides knowledge of a 'signature'
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Should `K` be capitalized or not? Whatever is right, we should one consistent.

File: draft-sullivan-cfrg-voprf.md [hunk 8]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
...
-- Verifiable: V must only complete execution of the protocol if it asserts that
-  P used its secret key k, associated with public key Y = kG, in execution.
 - Oblivious: P must learn nothing about V's input, and V must learn nothing
   about P's private key.
 - Unlinkable: If V reveals x to P, P cannot link x to the protocol instance in
   which y = F(k, x) was computed.
 
-# VOPRF Protocol {#protocol}
+Optionally, for any protocol that satisfies the above properties, there is a
+separate security property:
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/separate/additional?

File: draft-sullivan-cfrg-voprf.md [hunk 9]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
...
+
+- Verifiable: V must only complete execution of the protocol if it asserts that
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Is this the best way to phrase this particular property? Must V actually verify to achieve this? I'm just not sure. @grittygrease?

File: draft-sullivan-cfrg-voprf.md [hunk 10]
```diff
@@ -221,41 +262,59 @@ A prover P and verifier V engage in the following protocol given input m.
...
-L be the security parameter. Let k be the prover's (P) secret key, and Y = kG be
-its corresponding public key for some generator G taken from the group GG. Let x
-be the verifier's (V) input to the VOPRF protocol. (Commonly, it is a random
-L-bit string, though this is not required.) VOPRF begins with V randomly
-blinding its input for the signer. The latter then applies its secret key to the
+SHA256. All hash functions in the protocol are modelled as random oracles. Let L
+be the security parameter. Let k be the prover's (P) secret key, and Y = kG be
+its corresponding 'public key' for some generator G taken from the group GG.
+This public key is also referred to as a commitment to the key k. Let x be the
+verifier's (V) input to the VOPRF protocol. (Commonly, it is a random L-bit
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/VOPRF/OPRF

File: draft-sullivan-cfrg-voprf.md [hunk 11]
```diff
@@ -268,47 +327,73 @@ illustrated below.
...
+- VOPRF_Setup(l): Generate am integer k of sufficient bit-length l and output
+  (k, (G,Y)) where Y = kG for some generator G in GG.
+- VOPRF_Sign(k,(G,Y),M): Sign input M using secret key k to produce Z. Generate
+  a NIZK proof D = DLEQ_Generate(k,G,Y,M,Z), and output (Z, D).
+- VOPRF_Unblind(r,G,Y,M,(Z,D)): Unblind blinded signature Z with blind r,
   yielding N. Output N if 1 = DLEQ_Verify(G,Y,M,Z,D). Otherwise, output "error".
-- VOPRF_Finalize(N): Finalize N to produce PRF output F(k, x).
+
+When referring explicitly to VOPRF execution, we replace 'OPRF' in all method
+names with 'VOPRF'.
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Possibly add: "This means the VOPRF computation is finalized with 'OPRF_Finalize'."

File: draft-sullivan-cfrg-voprf.md [hunk 12]
```diff
@@ -322,30 +407,161 @@ of FF_p (elements u^2 where u is an element of FF_p) is cyclic, and we can pick
...
+1. P samples a uniformly random key k <- {0,1}^l for sufficient length l, and
+   interprets it as an integer.
+2. V computes X = H_1(x) and a random element r (blinding factor) from GF(p),
    and computes M = rX.
-2. V sends M to P.
-3. P computes Z = kM = rkX, and D = DLEQ_Generate(G,k,Y,M,Z).
-4. P sends (Z, D) to V.
-5. V ensures that 1 = DLEQ_Verify(G,Y,M,Z,D). If not, V outputs an error.
+3. V sends M to P.
+4. P computes Z = kM = rkX.
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Should we always multiply Z by the curve cofactor, to be safe?

File: draft-sullivan-cfrg-voprf.md [hunk 13]
```diff
@@ -322,30 +407,161 @@ of FF_p (elements u^2 where u is an element of FF_p) is cyclic, and we can pick
...
+ l: Some suitable choice of key-length (e.g. as described in {{NIST}}).
+
+Output:
+
+k: A key chosen from {0,1}^l and interpreted as an integer value.
+
+Steps:
+
+ 1. Sample k_bin <-$ {0,1}^l
+ 2. Output k <- (int)k_bin
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Can we move this to a specific routine that converts a binary string to an integer of fixed length? Relatedly, what is the width of `k`?

File: draft-sullivan-cfrg-voprf.md [hunk 14]
```diff
@@ -362,7 +578,7 @@ Steps:
 
  1.  r <-$ GF(p)
  2.  M := rH_1(x)
- 5.  Output (r, M)
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: 👍

File: draft-sullivan-cfrg-voprf.md [hunk 15]
```diff
@@ -591,13 +896,58 @@ way of "tagging" them. If all verifiers expect use of a certain private key,
...
+For some applications, it may be desirable for P to bind tokens to certain
+parameters, e.g., protocol versions, ciphersuites, etc. To accomplish this, P
+should use a distinct scalar for each parameter combination. Upon redemption of
+a token T from V, P can later verify that T was generated using the scalar
+associated with the corresponding parameters.
+
 # Acknowledgements
 
 This document resulted from the work of the Privacy Pass team {{PrivacyPass}}.
+The authors would also like to acknowledge the helpful conversations that they
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/that they had with/with

**Comment by alxdavids** [comment]
Thanks for your feedback! I've made the necessary fixes.

File: draft-sullivan-cfrg-voprf.md [hunk 16]
```diff
@@ -163,13 +163,13 @@ normative:
...
-evaluating the underlying PRF. This document specifies OPRF and VOPRF
-constructions instantiated within prime-order subgroups, including elliptic
-curves.
+that the server does not learn anything about the client's input during the
+evaluation. The client should also not learn anything about the server's secret
+PRF key. Optionally, OPRFs can also satisfy a notion 'verifiability' (VOPRF). In
+this setting, the client can verify that the server's output is indeed the
+result of evaluating the underlying PRF with just a public key. This document
+specifies OPRF and VOPRF constructions instantiated within prime-order
+subgroups, including elliptic curves.
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/subgroups/groups?

File: draft-sullivan-cfrg-voprf.md [hunk 17]
```diff
@@ -263,14 +265,14 @@ given input m.
 
 By the properties of RSA, s is clearly a valid signature for m. OPRF protocols
 can be used to provide a symmetric equivalent to blind signatures. Essentially
-the client learns PRF(k,x) for some input x of their choice, from a server that
-holds K. Since the security of an OPRF means that x is hidden in the
+the client learns y = PRF(k,x) for some input x of their choice, from a server that
+holds k. Since the security of an OPRF means that x is hidden in the
 interaction, then the client can later reveal x to the server along with
 PRF(k,x).
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: s/PRF(k,x)/y

File: draft-sullivan-cfrg-voprf.md [hunk 18]
```diff
@@ -485,6 +515,7 @@ Output:
 Steps:
 
  1. Z := kM
+ 2. Z <- qZ (q is elliptic curve cofactor)
```
Comment (chris-wood) [inline draft-sullivan-cfrg-voprf.md]: Several documents use `h` as the cofactor -- can we adopt that same terminology? Can we move it to a "definitions" section?

Comment (alxdavids) [reply draft-sullivan-cfrg-voprf.md]: I've mentioned the use of h in the opening steps of section {{#oprf}}

**Pull request merged**